### PR TITLE
Support returning how old the data is

### PR DESCRIPTION
**Pull request by gsabran**
## Changes
- add a column `modifiedAt ` in the cache
- change the `ApolloClient` protocol to return context alongside the result. So far the context (`GraphQLResultContext`) only contains the age of the returned result but could be later extended.
- extend `ApolloClient` to support existing usage where the handler doesn't care about this context

## Motivation
Reading from cache is a great way to reduce the number of network requests emitted, and cloud hosting costs 🤑. However developers usually want to make sure that the cache result is not too old, and without this information they are likely to fetch a fresh result _just in case_. This change makes it possible to know how old the returned result is.

```swift
apolloClient.fetchWithContext(
  query: query,
  cachePolicy: .returnCacheDataDontFetch,
  context: nil,
  queue: .main
) { result in
  switch result {
    case .success(let (data, context)):
      if context.resultAge > Date().milisecondsSince1970 - 3600 {
        // use fresh enough data
      } else {
        // fetch
      }
    default:
      // fetch
  }
}
```

## Tests 
I added test for the database migration, and for the context returned when reading from cache.

## Comments
I had to change a lot of files to keep track of the timestamps when passing around the object been built through the `GraphQLExecutor`.

### CODE REVIEW

File: Sources/Apollo/ApolloClientProtocol.swift [hunk 1]
```diff
@@ -83,7 +83,108 @@ public protocol ApolloClientProtocol: class {
...
   /// - Returns: An object that can be used to cancel an in progress subscription.
-  func subscribe<Subscription: GraphQLSubscription>(subscription: Subscription,
+  func subscribeWithContext<Subscription: GraphQLSubscription>(subscription: Subscription,
                                                     queue: DispatchQueue,
-                                                    resultHandler: @escaping GraphQLResultHandler<Subscription.Data>) -> Cancellable
+                                                    resultHandler: @escaping GraphQLResultWithContextHandler<Subscription.Data>) -> Cancellable
+}
+
+// MARK: Extension for result handler
+public extension ApolloClientProtocol {
```
[the remark this answers is not in the document]
Comment (gsabran) [reply Sources/Apollo/ApolloClientProtocol.swift]: The context is not an input variable, but returned to the result handler. So people would have to change their implementation entirely right?

File: Sources/Apollo/ApolloStore.swift [hunk 2]
```diff
@@ -198,74 +202,84 @@ public final class ApolloStore {
...
+      selections: [GraphQLSelection],
+      onObjectWithKey key: CacheKey,
+      variables: GraphQLMap?,
+      accumulator: Accumulator
+    ) throws -> Promise<Accumulator.FinalResult> {
+      return loadObject(forKey: key).flatMap { (object, lastModifiedAt) in
         let executor = GraphQLExecutor { object, info in
           let value = object[info.cacheKeyForField]
-          return self.complete(value: value)
+          return self.complete(value: value, firstModifiedAt: lastModifiedAt)
```
[the remark this answers is not in the document]
Comment (gsabran) [reply Sources/Apollo/ApolloStore.swift]: yes... each row has a `lastModifiedAt` column. The result is the aggregation of several rows and returns `min([lastModifiedAt])` ie the date of the oldest modification out of all the row read.

File: Tests/ApolloCacheDependentTests/LoadQueryFromStoreTests.swift [hunk 3]
```diff
@@ -319,17 +319,90 @@ class LoadQueryFromStoreTests: XCTestCase {
...
+  func testResultContextWithDataFromYesterday() throws {
+    let yesterday = Date().addingTimeInterval(-86400)
+    let aYearAgo = Date().addingTimeInterval(-86400 * 365)
+    let initialRecords = RecordSet([
+      "QUERY_ROOT": (["hero": Reference(key: "hero")], yesterday),
+      "hero": (["__typename": "Droid", "name": "R2-D2"], yesterday),
+      "somethingElse": (["__typename": "Droid", "name": "R2-D3"], aYearAgo),
+    ])
+
+    testResultContextWhenLoadingHeroNameQueryWithAge(initialRecords: initialRecords, expectedResultAge: yesterday)
```
[the remark this answers is not in the document]
Comment (gsabran) [reply Tests/ApolloCacheDependentTests/LoadQueryFromStoreTests.swift]: This one is not read by the query.

File: Sources/Apollo/ApolloStore.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (designatednerd) [reply Sources/Apollo/ApolloStore.swift]: Maybe a name like `rowModifiedAt` or `rowLastModifiedAt` would make this a bit clearer?

**Comment by gsabran** [comment]
Thanks @designatednerd . Let me do another pass and try to use `Date` directly in the storage. This should simplify a lot of things.

File: Sources/Apollo/ApolloClientProtocol.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (gsabran) [reply Sources/Apollo/ApolloClientProtocol.swift]: Also, I have mixed feeling about whether or not it's best to have extensions on the protocol or leave it to the implementation: all my code that uses Apollo only knowns that it is dealing with an `ApolloClientProtocol` which implementation depends on whether I'm running unit test or not. I think it is the best practice to have good testing.
So there's quite a bit of value to have the default values set in the protocol.

File: Sources/Apollo/RecordSet.swift [hunk 4]
```diff
@@ -1,70 +1,98 @@
...
   @discardableResult public mutating func merge(record: Record) -> Set<CacheKey> {
-    if var oldRecord = storage.removeValue(forKey: record.key) {
+    if var oldRow = storage.removeValue(forKey: record.key) {
       var changedKeys: Set<CacheKey> = Set()
-      
+
       for (key, value) in record.fields {
-        if let oldValue = oldRecord.fields[key], equals(oldValue, value) {
+        if let oldValue = oldRow.record.fields[key], equals(oldValue, value) {
           continue
```
Comment (gsabran) [inline Sources/Apollo/RecordSet.swift]: @designatednerd this part is not quite accurate as if the new record matches the existing one, we want to still update the `modifiedAt` to now to signal that this value is up to date. What is the best way to update the new value and all its references?
This also applies to `SQLiteNormalizedCache`

File: Sources/Apollo/ApolloClientProtocol.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (gsabran) [reply Sources/Apollo/ApolloClientProtocol.swift]: I'm totally fine with it. I was trying to go for a change that would require no code change for most people.

File: Sources/Apollo/RecordSet.swift (hunk 4 above)
Comment (gsabran) [reply Sources/Apollo/RecordSet.swift]: Also, correct me if I'm wrong, I think that when using SQLite, this returns all the keys the first time it is called for a specific result as it calls into the in memory `RecordSet` that would be empty at this point: https://github.com/apollographql/apollo-ios/blob/master/Sources/ApolloSQLite/SQLiteNormalizedCache.swift#L54

So it runs a db update for each key, which is a trade off against running a db read and preloading the existing entry in memory, but maybe not intentional.

This  makes the changes in this PR a bit harder to do as `changedFieldKeys` is not exactly what one would think.

**Comment by gsabran** [comment]
@designatednerd I think the terminology `modifiedAt` and `firstModifiedAt` need some changes as what I'm intending to track is the last time the entry was validated. ie if a response has the same value as the cached entry, this should update the field while there is no _modification_. Do you have a preference?

**Comment by designatednerd** [comment]
Gah, naming things is hard. 

`lastValidatedAt` doesn't quite have the right ring to it. Maybe `lastReceivedAt`? Because it hasn't been updated, but it has been received from the server and checked to make sure it's valid.

**Comment by designatednerd** [comment]
@gsabran sorry, but you're gonna have to do some rebasing 😬

**Comment by gsabran** [comment]
no worries, still to carve time to finish this..

**Comment by RolandasRazma** [comment]
what would be even more awesom if .returnCacheDataDontFetch would have parameter "to return for this amount and refetch after"

```swift
apolloClient.fetchWithContext(
  query: query,
  cachePolicy: .returnCacheDataDontFetch(for: 60),
  context: nil,
  queue: .main
) { result in
      // result is never older than 1 minute
}
```

**Comment by designatednerd** [comment]
@RolandasRazma That'd definitely be cool but I'd prefer to do that in a separate PR

**Comment by pgawlowski** [comment]
@gsabran  @designatednerd
What is the current state of this PR? Is there any chance I can help push this forward? This might be life-saver for my current issue.

**Comment by designatednerd** [comment]
I know @gsabran got sucked into some other stuff at work last time I talked to him about it - Gui, any update, or would you like some help?

**Comment by gsabran** [comment]
I'd love some help!

**Comment by designatednerd** [comment]
Gonna close this out in favor of #1498, in which @mordil took the work @gsabran started here and kept running with it. Please follow that PR for further updates!

**Pull request closed without merging**